prjxray/timing.py

```diff
@@ -0,0 +1,615 @@
+    """ All non-root TimingNode's are DownstreamNode's.
+
+    """
+
+    def propigate_delays(self, elements, math):
+        """ Propigates upstream delay elements to children of the tree.
+
+        Must call propigated_delays on all children of this node, and add this
+        node to elements.
+
+        Arguments
+        ---------
+        elements : list of TimingNode's
+            List of delay nodes between root of this tree and this node.
+        math : MathModel
+            Math model to use to compute delays
+        """
+
+
+class Outpin(TimingNode):
+    """ Represents a site output pin.
+
+    Outpin object is the root of the timing tree.  Once tree is built with
+    set_sink_wire and Wire.add_child methods, propigate_delays should be
+    invoked to estabilish model.
+
+    Arguments
+    ---------
+
+    resistance
+        Drive resistance in elmore delay model
+    delays
+        Intristic delays on output pin.
+
+    """
+
+    def __init__(self, resistance, delays):
+        self.resistance = resistance
+        self.delays = delays
+        self.sink_wire = None
+
+        self.downstream_cap = None
+        self.rc_delay = None
+
+    def set_sink_wire(self, wire):
+        """ Sets sink wire for this output pin.
+
+        An output pin always sinks to exactly 1 wire.
+
+        This method must be called prior to calling propigate_delays method
+        on this object.
+
+        Arguments
+        ---------
+        wire : Wire object
+            Sink wire for this output pin.
+
+        """
+        self.sink_wire = wire
+
+    def propigate_downstream_capacitance(self, math):
+        assert self.sink_wire is not None
+        self.downstream_cap = self.sink_wire.propigate_downstream_capacitance(
+            math)
+        self.rc_delay = math.multiply(self.downstream_cap, self.resistance)
+
+    def propigate_delays(self, math):
+        """ Propigate delays throughout tree using specified math model.
+
+        Must be called after elmore tree is estabilished.
+
+        Arguments
+        ---------
+        math : MathModel object
+            Math model used when doing timing computations.
+
+        """
+        self.propigate_downstream_capacitance(math)
+        self.sink_wire.propigate_delays([self], math)
+        self.rc_delay = math.multiply(self.resistance, self.downstream_cap)
+
+    def get_intrinsic_delays(self):
+        return self.delays
+
+    def get_rc_delay(self):
+        assert self.rc_delay is not None
+        return self.rc_delay
+
+    def get_downstream_cap(self):
+        assert self.downstream_cap is not None
+        return self.downstream_cap
+
+
+class Inpin(DownstreamNode):
+    """ Represents a site input pin.
+
+    Represents leaf of timing model.  Once model is connected and delays
+    are propigate (by calling Outpin,propigated_delays), get_delays will
+    correctly return the list of delay elements from the root to this leaf.
+
+    Arguments
+    ---------
+
```

Comment:
Types?

Reply:
The types here are intentionally omitted.  The types vary based on the math model in used.